### PR TITLE
Darren

### PR DESCRIPTION
**Pull request by darren2802**

### CODE REVIEW

File: lib/house.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
+require_relative './room'
+
+class House
+
+  attr_reader :price, :address, :rooms
+
+  def initialize(price, address)
+    @price = price.gsub('$','').to_i
```
Comment (aziobrow) [inline lib/house.rb]: Nice use of gsub! And of chaining methods.

File: lib/house.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    @address = address
+    @rooms = []
+  end
+
+  def add_room(room)
+    @rooms << room
```
Comment (aziobrow) [inline lib/house.rb]: Good job, very common pattern, so it's good you're getting familiar with it.

File: lib/house.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def rooms_from_category(category)
+    rooms_in_cat = []
+    @rooms.each { |room| rooms_in_cat << room if room.category == category }
```
Comment (aziobrow) [inline lib/house.rb]: This works, no problem (and nice job using single line syntax), but this is a great use case for a find_all/select.

File: lib/house.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    rooms_in_cat
+  end
+
+  def area
+    total_area = 0
+    @rooms.each { |room| total_area += room.area }
```
Comment (aziobrow) [inline lib/house.rb]: Same deal here--this totally works, but check out reduce!

https://apidock.com/ruby/Enumerable/reduce

File: lib/house.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    total_area
+  end
+
+  def price_per_square_foot
+    (@price / area.to_f).round(2)
+  end
+
+  def rooms_sorted_by_area
+    @rooms.sort_by { |room| room.area }
```
Comment (aziobrow) [inline lib/house.rb]: nicely done!

File: lib/house.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def area
+    total_area = 0
+    @rooms.each { |room| total_area += room.area }
+    total_area
+  end
+
+  def price_per_square_foot
+    (@price / area.to_f).round(2)
```
Comment (aziobrow) [inline lib/house.rb]: nice job

File: lib/house.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def rooms_sorted_by_area
+    @rooms.sort_by { |room| room.area }
+  end
+
+  def rooms_by_category
+    categories = []
+    @rooms.each do |room|
+      categories << room.category unless categories.include?(room.category)
```
Comment (aziobrow) [inline lib/house.rb]: This is a good use case for map, and ruby has a method called `uniq` that removes duplicate values. 

`@rooms.map { |room| room.category }.uniq` accomplishes the same thing as lines 38-41.

File: lib/house.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      rooms_by_cat[category] = []
+      @rooms.each do |room|
+        rooms_by_cat[category] << room if room.category == category
+      end
+    end
+
+    rooms_by_cat
+
+  end
+
```
Comment (aziobrow) [inline lib/house.rb]: Great job reasoning through the logical steps of this complex method and making it work!

File: lib/house.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    rooms_by_cat = Hash.new
+
+    categories.each do |category|
+      rooms_by_cat[category] = []
+      @rooms.each do |room|
+        rooms_by_cat[category] << room if room.category == category
+      end
+    end
+
```
Comment (aziobrow) [inline lib/house.rb]: These are pretty complicated ideas, so GREAT job figuring out how to make this work.  

For cleanup, again-- reduce is a great tool for the work you're doing. If you don't see it in your walkthrough, come see me, and I'll show you how it works.  

Also, remember, for lines 47-49: you have already done that work in the method for rooms_from_category, so just use that method instead of rewriting that logic.   Again, if this still doesn't make sense after the walkthrough, let me know, and I'll show you what I mean.

File: lib/room.rb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+class Room
+
+  attr_reader :category, :area
+
+  def initialize(category, length, width)
+    @category = category
+    @length = length
+    @width = width
+    @area = length * width
```
Comment (aziobrow) [inline lib/room.rb]: I see this as a method, not as an attribute, since it's using length and width to do some calculations.

File: test/house_test.rb [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  def setup
+    @house = House.new("$400000", "123 sugar lane")
+  end
+
+  def test_house_instantiated
+    assert_instance_of House, @house
+  end
+
+  def test_house_price
+    assert_equal 400000, @house.price
```
Comment (aziobrow) [inline test/house_test.rb]: Nice work following the specifics of the interaction pattern.

File: test/house_test.rb [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    assert_equal "123 sugar lane", @house.address
+  end
+
+  def test_house_no_rooms
+    assert_equal [], @house.rooms
+  end
+
+  def test_house_add_rooms
+    room_1 = Room.new(:bedroom, 10, 13)
+    assert_instance_of Room, room_1
```
Comment (aziobrow) [inline test/house_test.rb]: I like to put spaces between assertions in tests, to make it super clear where the assertions are happening and to clearly show the progression. It's a style thing, but it's nice.

Comment (aziobrow) [inline test/house_test.rb]: Since you tested that you could create a Room in the Room test, no need to retest it here in the House test.

File: test/house_test.rb [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    room_2 = Room.new(:bedroom, 11, 15)
+    assert_instance_of Room, room_2
+    @house.add_room(room_1)
+    @house.add_room(room_2)
+    assert_equal [room_1, room_2], @house.rooms
+  end
+
+  def test_add_4_rooms
```
Comment (aziobrow) [inline test/house_test.rb]: I think generally, if you can add 2 rooms, you can add 4 rooms, so this feels like a repetitive test to me.  If something's wrong in our code, it's usually the jump between having 0 rooms to having 1 room to having 2 rooms, and if it can do 2, it usually means it can do any number of them. Again, it doesn't hurt, but unnecessary to me.

File: test/house_test.rb [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    room_3 = Room.new(:living_room, 25, 15)
+    room_4 = Room.new(:basement, 30, 41)
+    @house.add_room(room_1)
+    @house.add_room(room_2)
+    @house.add_room(room_3)
+    @house.add_room(room_4)
+    assert_equal [room_1, room_2, room_3, room_4], @house.rooms
+  end
+
+  def test_add_4_rooms_get_metrics
```
Comment (aziobrow) [inline test/house_test.rb]: So we almost always want to keep tests so that they test only one method at a time, instead of lumping them together. I'm fine with multiple assertions in a test, so long as they're showing a progression of behavior within one method. But here, you're testing rooms_from_category as well as area, which I probably don't want to do. One test for area, a separate one for rooms_by_category.

File: test/house_test.rb [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    room_2 = Room.new(:bedroom, 11, 15)
+    room_3 = Room.new(:living_room, 25, 15)
+    room_4 = Room.new(:basement, 30, 41)
+    @house.add_room(room_1)
+    @house.add_room(room_2)
+    @house.add_room(room_3)
+    @house.add_room(room_4)
+    assert_equal [room_1, room_2, room_3, room_4], @house.rooms
+    assert_equal [room_1, room_2], @house.rooms_from_category(:bedroom)
+    assert_equal [room_4], @house.rooms_from_category(:basement)
```
Comment (aziobrow) [inline test/house_test.rb]: I like that you tested two different categories and made sure they were the right rooms. What happens if you pass in a category that doesn't exist? Think about edge case testing here.

File: test/house_test.rb [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    @house.add_room(room_1)
+    @house.add_room(room_2)
+    @house.add_room(room_3)
+    @house.add_room(room_4)
+    room_cats = {
+      bedroom: [room_1, room_2],
+      living_room: [room_3],
+      basement: [room_4]
+    }
+    assert_equal room_cats, @house.rooms_by_category
```
Comment (aziobrow) [inline test/house_test.rb]: Nice work on this-- as you continue to improve your skills, think about how to make tests more flexible (so that you can change things about test setup without breaking your test assertions), but this is super tough, so fantastic job figuring out a way to test this.
